code/obj/machinery/launcherloader.dm

```diff
@@ -411,37 +431,19 @@
 	New()
 		..()
 
-	attack_hand(var/mob/user as mob)
-		if (..(user))
-			return
-
-		var/dat = ""
-
-		dat += "<b>Print Amount:</b> <a href='?src=\ref[src];amount=1'>[src.print_amount]</a><BR>"
-
-		dat += "<BR><b>Available Destinations:</b><BR>"
-		for(var/I in destinations)
-			dat += "<b><A href='?src=\ref[src];print=[I]'>[I]</A></b><BR>"
-
-		dat += "<BR><b>Available Traders:</b><BR>"
-		for(var/datum/trader/T in shippingmarket.active_traders)
-			if (!T.hidden)
-				dat += "<b><A href='?src=\ref[src];print=[T.crate_tag]'>Sell to [T.name]</A></b><BR>"
-
-		dat += "<BR><b>Requisition Fulfillment:</b><BR>"
-		dat += "<b><A href='?src=\ref[src];print=["REQ-THIRDPARTY"]'>REQ-THIRDPARTY</A></b><BR>"
-		for(var/datum/req_contract/RC in shippingmarket.req_contracts)
-			dat += "<b><A href='?src=\ref[src];print=[RC.req_code]'>[RC.req_code]</A></b><BR>"
-
-		//dat += "<BR><b><A href='?src=\ref[src];add=1'>Add Tag</A></b>"
-
-		src.add_dialog(user)
-		// Attempting to diagnose an infinite window refresh I can't duplicate, reverting the display style back to plain HTML to see what results that gets me.
-		// Hooray for having a playerbase to test shit on
-		//user.Browse(dat, "title=Barcode Computer;window=bc_computer_[src];size=300x480")
-		user.Browse(dat, "title=Barcode Computer;window=bc_computer_[src];size=300x480")
-		onclose(user, "bc_computer_[src]")
-		return
+	ui_data(mob/user)
+		. = ..()
+		var/list/traders = new()
+		for (var/datum/trader/T in shippingmarket.active_traders)
+			if (T.hidden)
+				continue
+			traders += list(list("crate_tag" = T.crate_tag, "name" = T.name))
+		.["sections"] += list(list("title" = "Traders", "destinations" = traders))
+		var/list/req_codes = new()
+		req_codes += list(list("crate_tag" = "REQ-THIRDPARTY", "name" = "Third party"))
+		for (var/datum/req_contract/RC in shippingmarket.req_contracts)
+			req_codes += list(list("crate_tag" = RC.req_code, "name" = RC.name))
+		.["sections"] += list(list("title" = "Requisition contracts", "destinations" = req_codes))
```

Comment:
I feel like the information in here could be static data, and then manually refreshed when the active traders/contracts list changes?

Reply:
Hmm, how to detect when they change though? The only way I can think of is to set a static flag on the barcode computer class from the shipping market whenever it modifies the lists and then get ui_data to check the flag and call update_static_data. This seems kinda flimsy and jank, is there a better way?

Reply:
Hmm not sure at the moment, can't really look into it. However, doing this data computation every 0.9s is kind of a lot.

Reply:
I'll try and figure out something reasonable tomorrow

Reply:
Why not just 
```
modify_shipping_data() / modify_trader_actives()
...
for (barcode computer in either machine registry or by_type)
	computer.update_static_data()
```